TF-TRT 3D pooling

ralovich opened the pull request:
Allow conversion of AvgPool3D and MaxPool3D ops for TensorRT offload. TensorRT 6+ supports these operators. In medical imaging applications 3D convolutions are often paired with 3D pooling. Since https://github.com/tensorflow/tensorflow/commit/a2fa0c58dcd81307a3eac76ccc2b81a42b71583c 3D convolutions can be offloaded to TRT 6+ and this change allows offloading 3D pooling too. This PR allows to run 3D U-Nets through TF-TRT.

Commit:
Allow conversion of AvgPool3D and MaxPool3D ops for TensorRT offload

Commit:
Formatting according to Google C++ Style Guide

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2906,6 +2906,76 @@ Status ConvertConv3D(OpConverterParams* params) {
   return ConvertConv3DHelper(params, 1, /*is_conv3d_backprop_input=*/true);
 }
+
+Status ConvertPool3D(OpConverterParams* params) {
```
Thanks for the improvement! Could you help to add a test for this?

ralovich replied:
Added tests as requested.

ralovich replied:
Added tests as requested.

aaroey reviewed (changes requested):
Very nice! Once the test is added we're good to go.

Commit:
Add unit tests for TF-TRT offloading of 3D pooling ops

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -4310,6 +4310,216 @@ TEST_F(OpConverterTest, ConvertConv3D) {
+      // Basic - just 1x1 max pooling - input = output
+      TestParams{
+          /*input_dims=*/{1, 3, 3, 3},
+          /*input=*/{1, 2,  15,  3, 6,  -3, 22, 1, 88, 56, 36, 1,  1, 105,
```
nit: this input is repeated many times, can we make it a `const vector<float>` and reuse it for all occurrences?

ralovich replied:
Addressed in https://github.com/tensorflow/tensorflow/pull/32843/commits/ab48f4961cc0e3fb31085b6a888dd38cbc9e80fa .

aaroey commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -4310,6 +4310,216 @@ TEST_F(OpConverterTest, ConvertConv3D) {
+          /*data_format=*/"NDHWC",
+          /*is_max_pooling=*/false,
+          /*expected_output_dims=*/{1, 1, 1, 1},
+          /*expected_output=*/{17.1851851851852}},
```
Can we make the input divisible by 27? Otherwise this can make the test fail easily due to numerical issues. Alternatively we can allow errors below instead of `ElementsAreArray` below.

ralovich replied:
Addressed in https://github.com/tensorflow/tensorflow/pull/32843/commits/ab48f4961cc0e3fb31085b6a888dd38cbc9e80fa .

Commit:
Address review findings in unit-tests

Factor out constant array and reuse it. Make sure the average of the
test tensors is integer.

aaroey reviewed (dismissed):
Thanks for the improvement!

pooyadavoodi commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2906,6 +2906,95 @@ Status ConvertConv3D(OpConverterParams* params) {
+  const nvinfer1::Dims3 ksize(tf_kernel[d_index], tf_kernel[h_index],
+                              tf_kernel[w_index]);
+
+  nvinfer1::INetworkDefinition* network = params->converter->network();
```
To be consistent with other converters, please do this: `params->converter->network()->addPoolingNd`

ralovich replied:
Addressed in https://github.com/tensorflow/tensorflow/pull/32843/commits/8c8afe88694ac40d6c64f3d999dc29e8d7a23da7 .

Commit:
Remove debug leftover

pooyadavoodi commented:
Thanks @ralovich for the PR.
I only had a small comment.

ralovich commented:
Dear https://github.com/aaroey and https://github.com/pooyadavoodi, all points should be addressed now. Could you please verify?
